pkg/services/sqlstore/migrations/ualert/tables.go

```diff
@@ -241,6 +241,16 @@ func AddAlertRuleMigrations(mg *migrator.Migrator, defaultIntervalSeconds int64)
 			Cols: []string{"org_id", "dashboard_uid", "panel_id"},
 		},
 	))
+
+	mg.AddMigration("add rule_group_idx column to alert_rule", migrator.NewAddColumnMigration(
```

Comment:
Should we also add an unique index to make sure that rules in the same Rule Group cannot have the same `rule_group_idx`?

Reply:
Isn't it possible that two transactions at the same time could cause two rows have the same index? For example, a `POST` request that adds a new row and a second `POST` request that changes the order of rules in the Rule Group?

For example, the first `POST` request changes Rules A and B around so the order is now B, A, C, D while the second `POST` request changes Rules B and C around so the order is now A, C, B, D. The first transaction will change the index for Rules A and B to 2 and 1, while the second transaction will change the order of Rules B and C to 3 and 2. This means the final outcome if the first transaction commits and then the second transaction would be:

    Tx 1        Tx 2        Outcome
	Rule A: 2   Rule B: 3   Rule A: 2
	Rule B: 1   Rule C: 2   Rule B: 3
	                        Rule C: 2
	                        Rule D: 4

Reply:
Isn't it possible that two transactions at the same time could cause two rows have the same index?

For example, the first `POST` request changes Rules A and B around so the order is now B, A, C, D while the second `POST` request changes Rules B and C around so the order is now A, C, B, D. The first transaction will change the index for Rules A and B to 2 and 1, while the second transaction will change the order of Rules B and C to 3 and 2. This means the final outcome if the first transaction commits and then the second transaction would be:

    Tx 1        Tx 2        Outcome
	Rule A: 2   Rule B: 3   Rule A: 2
	Rule B: 1   Rule C: 2   Rule B: 3
	                        Rule C: 2
	                        Rule D: 4

Reply:
Yes, you're right. The problem is that we do not update _all_ rules in the group. Every transaction should update every rule in the group.

Reply:
I see a draft PR below this for optimistic locking, do we plan to fix it there?

Reply:
PR https://github.com/grafana/grafana/pull/50274 is a prerequisite for changes in this one.

Reply:
Ok, I addressed the comment. Now we will update (just version, and update timestamp) all rules in all affected groups (there can be many groups when the user moves alert from one group to another). Effectively, this will mean that there cannot be any two concurrent change of rules in the same groups.